controllers/upgrade.go

```diff
@@ -777,57 +840,193 @@ func (r *RollingUpgradeContext) ClusterBallooning(batchSize int) (bool, int) {
 	return false, batchSize
 }
 
+const (
+	// Annotation key to track nodes cordoned by the early cordon feature
+	EarlyCordonAnnotationKey   = "rollingupgrade.keikoproj.io/early-cordoned-by"
+	EarlyCordonAnnotationValue = "upgrade-manager"
+)
+
+// CordonUncordonAllNodes cordons or uncordons all drifted nodes in the scaling group
+// When cordonNode is true we cordon all drifted nodes that aren't already cordoned
+// When cordonNode is false we uncordon all nodes that were previously cordoned by upgrade-manager
 func (r *RollingUpgradeContext) CordonUncordonAllNodes(cordonNode bool) (bool, error) {
 	scalingGroup := awsprovider.SelectScalingGroup(r.RollingUpgrade.ScalingGroupName(), r.Cloud.ScalingGroups)
-	var instanceIDs []string
-	var err error
 
-	if cordonNode {
-		instanceIDs, err = r.Cloud.AmazonClientSet.DescribeInstancesWithoutTagValue(instanceStateTagKey, earlyCordonedTagValue)
-		if err != nil {
-			r.Error(err, "failed to describe instances for early-cordoning", "name", r.RollingUpgrade.NamespacedName())
-			return false, errors.Wrap(err, "failed to describe instances for early-cordoning")
+	var processedCount, errorCount int
+
+	for _, instance := range scalingGroup.Instances {
+		// Skip terminating instances
+		if common.ContainsEqualFold(awsprovider.TerminatingInstanceStates, aws.StringValue(instance.LifecycleState)) {
+			continue
+		}
+
+		// Skip non-drifted instances
+		if !r.IsInstanceDrifted(instance) {
```

Comment:
Ideally, non drifted instance should never be considered for earlyCordon. 

But in case a non-drifted instance was early cordoned (when the launch template is switched back and forth. instance-manager could do this, if the latest launch template has issues.) after checking for the annotation, we should be uncordoning it.

Reply:
Sounds good, I added an extra conditional to ensure that we do not consider non-drifted instances when cordonNode is true.

For the case you mentioned, I believe we handle that here:
https://github.com/keikoproj/upgrade-manager/pull/518/files#diff-077ba5eb59266ac27713b3442231daf3287b93fa4b6683c7a58aea43c0e73f12R890-R900